lewis/adapters/stream.py

```
@@ -86,7 +84,44 @@ def handle_accept(self):
             StreamHandler(sock, self.target)
 
 
-class Cmd(object):
+class CanNotProcessException(Exception):
+    pass
```

This sort of goes back to #135... I can see how it is convenient to make ad hoc exception types where you need them to catch only that specific exception type, but is this good practice in general?

---

It goes a bit with the comment above, I tried with different return values first, but somehow returning a mix of error code (what match essentially is) and `reply`, which is the actual result didn't really appeal to me - especially since I'm not aware we're using this "pattern" elsewhere.

On the other hand this is kind of "misusing exceptions for flow control", but only because there is no other use of `process_request` (at the moment). Otherwise I think raising an exception would be the right thing to do here - "I tried processing the request, I wasn't able to do it, maybe someone else is". I also used a special type, because that is an exception that is expected and we know how to deal with it. Other exceptions may result from actually executing the function(s) and for those we actually do not know and we catch them very broadly to forward them to the `handle_error` method.

A completely different way would be to have an additional method `can_process(request)` which just checks the regex and returns True or False. In case of actually processing the command, the request would then be run against the regex again to extract the groups. Or the `process_request` method takes a regex match object, but somehow I don't think having those in such a general method signature is particularly nice either...

I first thought this could be some sort of "Chain of Responsibility" pattern, but I think that would require a new type that is returned from the function that makes it possible to distinguish between "this is the original request" and "this is the result of the processing".

Any other ideas how to handle this?

---

Thought about this some more. It's a bit awkward no matter how you do it. But I'm still leaning towards the tuple return.

Introducing `can_process` would give us a way to check before calling `process_request`, but then you have to use the same match in both places. And I could just call `process_request` with garbage values anyway... so it would still need to throw an exception on no match. So, in effect, we haven't simplified anything and only added code duplication and complexity.

Chain of responsibility seems like overkill for this simple situation. We could consider a simplified version, like what the `re` module does: Take parameters normally but return a `match` object that may or may not have results, and contains meta information about the nature and validity of results.

But what would our return object need to look like / what values would it contain? It would contain only two things: 

1. A boolean to indicate whether the request was accepted/processed, and 
2. A return value from processing the request (default `None` if above `False`). 

So: `(False, None)` or `(True, ret_value)`. Why create a new type for this if tuple fits the bill?

---

From my point of view "And I could just call process_request with garbage values anyway... so it would still need to throw an exception on no match." is the key argument here. This is how I would write the function if someone told me "Write a function that processes something and return the result". I would make it raise an exception, I'm just not entirely sure which one...